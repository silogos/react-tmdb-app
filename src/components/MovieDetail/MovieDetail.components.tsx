import { NavLink } from "react-router";
import { ChevronLeftIcon } from "@heroicons/react/16/solid";
import {
  CalendarDaysIcon,
  ClockIcon,
  StarIcon,
} from "@heroicons/react/16/solid";

import { getReleaseYear } from "@/utils/Movie.utils";

import {
  getRatingText,
  formatRunTime,
  formatReleaseDate,
} from "@/utils/Movie.utils";
import Banner from "../Banner";
import Poster from "../Poster";
import {
  MovieDetailHeaderProps,
  MovieDetailHeroProps,
} from "./MovieDetail.type";

export function MovieDetailHeader({ data }: MovieDetailHeaderProps) {
  return (
    <header className="overflow-hidden mb-2 bg-gray-300 bg-opacity-20">
      <div className="mx-auto max-w-screen-xl px-8 py-4 text-gray-400">
        <div className="flex flex-row gap-4 items-stretch">
          <Poster poster={data.poster_path} className="w-20" />

          <div className="flex flex-col justify-evenly">
            <h1 className="text-left text-white text-xl lg:text-4xl">
              {data.title} ({getReleaseYear(data.release_date)})
            </h1>

            <NavLink
              to={".."}
              className="cursor-pointer hover:underline flex flex-row items-center text-base lg:text-lg"
            >
              <ChevronLeftIcon className="size-6 inline" />
              <span>Back to list</span>
            </NavLink>
          </div>
        </div>
      </div>
    </header>
  );
}

export function MovieDetailHero({ data }: MovieDetailHeroProps) {
  return (
    <section id="hero" className="relative overflow-hidden">
      <div className="mx-auto max-w-screen-xl px-8 py-8">
        <div className="flex flex-col lg:flex-row gap-10">
          <Banner className="lg:h-96" src={data.backdrop_path} />

          <div className="py-4">
            <h1 className="text-left text-white text-xl lg:text-4xl mb-4">
              {data.title} ({getReleaseYear(data.release_date)})
            </h1>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base">
              {data.genres.map((genre) => (
                <div
                  key={genre.id}
                  className="border border-gray-300 text-sm px-2 py-1 rounded-full "
                >
                  {genre.name}
                </div>
              ))}
            </div>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base">
              <span className="text-left font-bold">
                <CalendarDaysIcon className="size-5 inline mr-1 mb-1" />
                {formatReleaseDate(data.release_date)}
              </span>
              <span>•</span>
              <span className="text-left font-bold">
                <StarIcon className="size-5 inline mr-1 mb-1" />
                {getRatingText(data.vote_average)}
              </span>
              <span>•</span>
              <span className="text-left font-bold">
                <ClockIcon className="size-5 inline mr-1" />
                {formatRunTime(data.runtime)}
              </span>
            </div>

            <p className="text-left text-gray-300 text-base my-4">
              <i>{data.tagline}</i>
            </p>

            <span className="text-left text-white text-lg font-bold mb-1">
              Overview
            </span>
            <p className="text-left text-gray-300 text-base">{data.overview}</p>
          </div>
        </div>
      </div>
    </section>
  );
}
