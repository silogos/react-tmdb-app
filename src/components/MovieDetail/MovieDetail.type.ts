import { MovieDetail } from "@/types/Movie.types";

export type MovieDetailHeaderProps = {
  data: MovieDetail;
};
export type MovieDetailHeroProps = {
  data: MovieDetail;
};
