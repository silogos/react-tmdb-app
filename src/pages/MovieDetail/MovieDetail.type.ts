import { MovieDetail } from "@/types/Movie.types";

export type MovieDetailHeroProps = {
  data: MovieDetail;
};
