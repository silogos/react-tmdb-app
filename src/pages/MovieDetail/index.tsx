export { default } from "./MovieDetail.page";
