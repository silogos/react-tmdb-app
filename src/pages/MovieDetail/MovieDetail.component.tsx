import { NavLink } from "react-router";
import { ChevronLeftIcon } from "@heroicons/react/16/solid";
import {
  CalendarDaysIcon,
  ClockIcon,
  StarIcon,
} from "@heroicons/react/16/solid";

import {
  getRatingText,
  formatRunTime,
  formatReleaseDate,
  getReleaseYear,
} from "@/utils/Movie.utils";

import { MovieDetailHeroProps } from "./MovieDetail.type";
import Banner from "@/components/Banner";

export function Header() {
  return (
    <header className="overflow-hidden mb-2 bg-gray-300 bg-opacity-20">
      <div className="mx-auto max-w-screen-xl px-8 py-4 text-gray-400">
        <div className="flex flex-row gap-4 items-stretch">
          <div className="flex flex-row justify-evenly gap-2 items-center">
            <NavLink
              to={".."}
              className="cursor-pointer hover:underline flex flex-row items-center text-sm"
            >
              <ChevronLeftIcon className="size-6 inline" />
              Back to list
            </NavLink>
          </div>
        </div>
      </div>
    </header>
  );
}

export function Hero({ data }: MovieDetailHeroProps) {
  return (
    <section id="hero" className="relative overflow-hidden">
      <div className="mx-auto max-w-screen-xl px-8 py-2 sm:py-4 lg:py-8">
        <div className="flex flex-col lg:flex-row gap-2 lg:gap-10">
          <Banner className="lg:h-96" src={data.backdrop_path} />

          <div className="py-4">
            <h1 className="text-left text-white text-xl sm:text-2xl lg:text-4xl mb-4">
              {data.title} ({getReleaseYear(data.release_date)})
            </h1>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base">
              {data.genres.map((genre) => (
                <div
                  key={genre.id}
                  className="border border-gray-300 text-xs lg:text-sm px-2 py-1 rounded-full "
                >
                  {genre.name}
                </div>
              ))}
            </div>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-xs sm:text-sm lg:text-base">
              <span className="text-left font-bold text-nowrap">
                <CalendarDaysIcon className="size-5 inline mr-1 mb-1" />
                {formatReleaseDate(data.release_date)}
              </span>
              <span>•</span>
              <span className="text-left font-bold text-nowrap">
                <StarIcon className="size-5 inline mr-1 mb-1" />
                {getRatingText(data.vote_average)}
              </span>
              <span>•</span>
              <span className="text-left font-bold text-nowrap">
                <ClockIcon className="size-5 inline mr-1" />
                {formatRunTime(data.runtime)}
              </span>
            </div>

            <p className="text-left text-gray-300 text-xs sm:text-sm lg:text-base my-4">
              <i>{data.tagline}</i>
            </p>

            <p className="text-left text-white text-sm sm:text-base lg:text-lg font-bold mb-1">
              Overview
            </p>
            <p className="text-left text-gray-300 text-xs sm:text-sm lg:text-base">
              {data.overview}
            </p>
          </div>
        </div>
      </div>
    </section>
  );
}

export function HeroShimmering() {
  return (
    <div className="animate-pulse mx-auto max-w-screen-xl px-8 py-2 sm:py-4 lg:py-8">
      <div className="flex flex-col lg:flex-row gap-8 lg:gap-10">
        <div
          className={`relative lg:h-96 aspect-video bg-gray-700 bg-opacity-90 rounded-lg overflow-hidden`}
        ></div>
        <div className="flex-1 flex flex-col gap-4">
          <div className="w-full h-8 lg:h-10 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="flex flex-row gap-4">
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
          </div>
          <div className="w-3/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-2/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-3/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-2/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-1/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
        </div>
      </div>
    </div>
  );
}
