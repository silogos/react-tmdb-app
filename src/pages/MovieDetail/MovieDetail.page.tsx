import CreditList from "@/components/CreditList";
import useMovieDetail from "@/hooks/useMovieDetail";
import useCredits from "@/hooks/useCredits";

import { Header, Hero, HeroShimmering } from "./MovieDetail.component";
import useBackgroundImage from "@/hooks/useBackgroundImage";
import { DisplayError } from "@/components/Displays";

function MovieDetailPage() {
  const { data, isLoading, isError, getData } = useMovieDetail();
  const { casts, crews } = useCredits();
  useBackgroundImage(data?.backdrop_path);

  return (
    <>
      <div className="relative w-full min-h-screen">
        <Header />

        {isError && (
          <div className="mx-auto max-w-screen-xl p-8">
            <DisplayError onReload={getData} />
          </div>
        )}

        {isLoading && <HeroShimmering />}

        {!isLoading && data && (
          <>
            <Hero data={data} />
            <CreditList title="Cast" credits={casts} />
            <CreditList title="Crew" credits={crews} />
          </>
        )}
      </div>
    </>
  );
}

export default MovieDetailPage;
