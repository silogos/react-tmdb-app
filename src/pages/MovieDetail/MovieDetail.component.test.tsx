import { describe, it, expect, vi } from "vitest";
import { Header, Hero, HeroShimmering } from "./MovieDetail.component";
import { render } from "@testing-library/react";
import { MovieDetail } from "@/types/Movie.types";
import { BrowserRouter } from "react-router";

const dummyData: MovieDetail = {
  adult: false,
  backdrop_path: "/b85bJfrTOSJ7M5Ox0yp4lxIxdG1.jpg",
  belongs_to_collection: {
    id: 720879,
    name: "Sonic the Hedgehog Collection",
    poster_path: "/fwFWhYXj8wY6gFACtecJbg229FI.jpg",
    backdrop_path: "/l5CIAdxVhhaUD3DaS4lP4AR2so9.jpg",
  },
  budget: 122000000,
  genres: [
    {
      id: 28,
      name: "Action",
    },
    {
      id: 878,
      name: "Science Fiction",
    },
    {
      id: 35,
      name: "Comedy",
    },
    {
      id: 10751,
      name: "Family",
    },
  ],
  homepage: "https://www.sonicthehedgehogmovie.com",
  id: 939243,
  imdb_id: "tt18259086",
  origin_country: ["US"],
  original_language: "en",
  original_title: "Sonic the Hedgehog 3",
  overview:
    "Sonic, Knuckles, and Tails reunite against a powerful new adversary, Shadow, a mysterious villain with powers unlike anything they have faced before. With their abilities outmatched in every way, Team Sonic must seek out an unlikely alliance in hopes of stopping Shadow and protecting the planet.",
  popularity: 3560.42,
  poster_path: "/d8Ryb8AunYAuycVKDp5HpdWPKgC.jpg",
  production_companies: [
    {
      id: 4,
      logo_path: "/gz66EfNoYPqHTYI4q9UEN4CbHRc.png",
      name: "Paramount Pictures",
      origin_country: "US",
    },
    {
      id: 333,
      logo_path: "/5xUJfzPZ8jWJUDzYtIeuPO4qPIa.png",
      name: "Original Film",
      origin_country: "US",
    },
    {
      id: 77884,
      logo_path: "/dP2lxVNctD5Cried0IWVqgrO2o9.png",
      name: "Marza Animation Planet",
      origin_country: "JP",
    },
    {
      id: 113750,
      logo_path: "/A3QVZ9Ah0yI2d2GiXUFpdlbTgyr.png",
      name: "SEGA",
      origin_country: "JP",
    },
    {
      id: 10644,
      logo_path: "/ocLZIdYJBppuCt1rhYEb2jbpt5F.png",
      name: "Blur Studio",
      origin_country: "US",
    },
    {
      id: 168701,
      logo_path: "/vWdZFT4V64CCv12D10m44duQjyg.png",
      name: "SEGA of America",
      origin_country: "US",
    },
  ],
  production_countries: [
    {
      iso_3166_1: "JP",
      name: "Japan",
    },
    {
      iso_3166_1: "US",
      name: "United States of America",
    },
  ],
  release_date: "2024-12-19",
  revenue: 422398000,
  runtime: 110,
  spoken_languages: [
    {
      english_name: "English",
      iso_639_1: "en",
      name: "English",
    },
    {
      english_name: "French",
      iso_639_1: "fr",
      name: "Français",
    },
    {
      english_name: "German",
      iso_639_1: "de",
      name: "Deutsch",
    },
    {
      english_name: "Japanese",
      iso_639_1: "ja",
      name: "日本語",
    },
    {
      english_name: "Letzeburgesch",
      iso_639_1: "lb",
      name: "",
    },
    {
      english_name: "Malagasy",
      iso_639_1: "mg",
      name: "",
    },
    {
      english_name: "Spanish",
      iso_639_1: "es",
      name: "Español",
    },
  ],
  status: "Released",
  tagline: "Try to keep up.",
  title: "Sonic the Hedgehog 3",
  video: false,
  vote_average: 7.597,
  vote_count: 725,
};

vi.mock("@/components/Banner/Banner.component");

describe("MovieDetail.components", () => {
  it("renders Header correctly", () => {
    const component = render(
      <BrowserRouter>
        <Header />
      </BrowserRouter>
    );

    expect(component.container).toMatchSnapshot();
  });

  it("renders Hero correctly", () => {
    const component = render(<Hero data={dummyData} />);

    expect(component.container).toMatchSnapshot();
  });

  it("renders HeroShimmering correctly", () => {
    const component = render(<HeroShimmering />);

    expect(component.container).toMatchSnapshot();
  });
});
