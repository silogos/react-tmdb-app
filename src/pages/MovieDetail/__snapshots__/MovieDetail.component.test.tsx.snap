// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieDetail.components > renders Header correctly 1`] = `
<div>
  <header
    class="overflow-hidden mb-2 bg-gray-300 bg-opacity-20"
  >
    <div
      class="mx-auto max-w-screen-xl px-8 py-4 text-gray-400"
    >
      <div
        class="flex flex-row gap-4 items-stretch"
      >
        <div
          class="flex flex-row justify-evenly gap-2 items-center"
        >
          <a
            aria-current="page"
            class="cursor-pointer hover:underline flex flex-row items-center text-sm active"
            data-discover="true"
            href="/"
          >
            <svg
              aria-hidden="true"
              class="size-6 inline"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M9.78 4.22a.75.75 0 0 1 0 1.06L7.06 8l2.72 2.72a.75.75 0 1 1-1.06 1.06L5.47 8.53a.75.75 0 0 1 0-1.06l3.25-3.25a.75.75 0 0 1 1.06 0Z"
                fill-rule="evenodd"
              />
            </svg>
            Back to list
          </a>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`MovieDetail.components > renders Hero correctly 1`] = `
<div>
  <section
    class="relative overflow-hidden"
    id="hero"
  >
    <div
      class="mx-auto max-w-screen-xl px-8 py-2 sm:py-4 lg:py-8"
    >
      <div
        class="flex flex-col lg:flex-row gap-2 lg:gap-10"
      >
        <div
          data-testid="Banner"
        />
        <div
          class="py-4"
        >
          <h1
            class="text-left text-white text-xl sm:text-2xl lg:text-4xl mb-4"
          >
            Sonic the Hedgehog 3
             (
            2024
            )
          </h1>
          <div
            class="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base"
          >
            <div
              class="border border-gray-300 text-xs lg:text-sm px-2 py-1 rounded-full "
            >
              Action
            </div>
            <div
              class="border border-gray-300 text-xs lg:text-sm px-2 py-1 rounded-full "
            >
              Science Fiction
            </div>
            <div
              class="border border-gray-300 text-xs lg:text-sm px-2 py-1 rounded-full "
            >
              Comedy
            </div>
            <div
              class="border border-gray-300 text-xs lg:text-sm px-2 py-1 rounded-full "
            >
              Family
            </div>
          </div>
          <div
            class="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-xs sm:text-sm lg:text-base"
          >
            <span
              class="text-left font-bold text-nowrap"
            >
              <svg
                aria-hidden="true"
                class="size-5 inline mr-1 mb-1"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.75 7.5a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5ZM5 10.25a.75.75 0 1 1 1.5 0 .75.75 0 0 1-1.5 0ZM10.25 7.5a.75.75 0 1 0 0 1.5.75.75 0 0 0 0-1.5ZM7.25 8.25a.75.75 0 1 1 1.5 0 .75.75 0 0 1-1.5 0ZM8 9.5A.75.75 0 1 0 8 11a.75.75 0 0 0 0-1.5Z"
                />
                <path
                  clip-rule="evenodd"
                  d="M4.75 1a.75.75 0 0 0-.75.75V3a2 2 0 0 0-2 2v7a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2V1.75a.75.75 0 0 0-1.5 0V3h-5V1.75A.75.75 0 0 0 4.75 1ZM3.5 7a1 1 0 0 1 1-1h7a1 1 0 0 1 1 1v4.5a1 1 0 0 1-1 1h-7a1 1 0 0 1-1-1V7Z"
                  fill-rule="evenodd"
                />
              </svg>
              12/19/2024
            </span>
            <span>
              •
            </span>
            <span
              class="text-left font-bold text-nowrap"
            >
              <svg
                aria-hidden="true"
                class="size-5 inline mr-1 mb-1"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M8 1.75a.75.75 0 0 1 .692.462l1.41 3.393 3.664.293a.75.75 0 0 1 .428 1.317l-2.791 2.39.853 3.575a.75.75 0 0 1-1.12.814L7.998 12.08l-3.135 1.915a.75.75 0 0 1-1.12-.814l.852-3.574-2.79-2.39a.75.75 0 0 1 .427-1.318l3.663-.293 1.41-3.393A.75.75 0 0 1 8 1.75Z"
                  fill-rule="evenodd"
                />
              </svg>
              7.6/10
            </span>
            <span>
              •
            </span>
            <span
              class="text-left font-bold text-nowrap"
            >
              <svg
                aria-hidden="true"
                class="size-5 inline mr-1"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M1 8a7 7 0 1 1 14 0A7 7 0 0 1 1 8Zm7.75-4.25a.75.75 0 0 0-1.5 0V8c0 .414.336.75.75.75h3.25a.75.75 0 0 0 0-1.5h-2.5v-3.5Z"
                  fill-rule="evenodd"
                />
              </svg>
              1h 50m
            </span>
          </div>
          <p
            class="text-left text-gray-300 text-xs sm:text-sm lg:text-base my-4"
          >
            <i>
              Try to keep up.
            </i>
          </p>
          <p
            class="text-left text-white text-sm sm:text-base lg:text-lg font-bold mb-1"
          >
            Overview
          </p>
          <p
            class="text-left text-gray-300 text-xs sm:text-sm lg:text-base"
          >
            Sonic, Knuckles, and Tails reunite against a powerful new adversary, Shadow, a mysterious villain with powers unlike anything they have faced before. With their abilities outmatched in every way, Team Sonic must seek out an unlikely alliance in hopes of stopping Shadow and protecting the planet.
          </p>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`MovieDetail.components > renders HeroShimmering correctly 1`] = `
<div>
  <div
    class="animate-pulse mx-auto max-w-screen-xl px-8 py-2 sm:py-4 lg:py-8"
  >
    <div
      class="flex flex-col lg:flex-row gap-8 lg:gap-10"
    >
      <div
        class="relative lg:h-96 aspect-video bg-gray-700 bg-opacity-90 rounded-lg overflow-hidden"
      />
      <div
        class="flex-1 flex flex-col gap-4"
      >
        <div
          class="w-full h-8 lg:h-10 rounded-lg bg-gray-700 bg-opacity-90"
        />
        <div
          class="flex flex-row gap-4"
        >
          <div
            class="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90"
          />
          <div
            class="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90"
          />
          <div
            class="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90"
          />
          <div
            class="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90"
          />
        </div>
        <div
          class="w-3/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90"
        />
        <div
          class="w-2/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90"
        />
        <div
          class="w-3/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90"
        />
        <div
          class="w-2/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90"
        />
        <div
          class="w-1/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90"
        />
      </div>
    </div>
  </div>
</div>
`;
